### PR TITLE
[Validator] Fix precision issue regarding floats and DivisibleBy constraint

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Alternative to #28212

### CODE REVIEW

**Comment by apfelbox** [comment]
Looks great. 👍 
I am uncomfortable with directly comparing floats, though. Do you think using the "standard epsilon" approach might catch more edge cases?

File: src/Symfony/Component/Validator/Constraints/DivisibleByValidator.php [hunk 1]
```diff
@@ -23,7 +23,11 @@ class DivisibleByValidator extends AbstractComparisonValidator
      */
     protected function compareValues($value1, $value2)
     {
-        return (float) 0 === fmod($value1, $value2);
+        if (!$value1 = fmod($value1, $value2)) {
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/DivisibleByValidator.php]: it is, so that the check just below uses the rest returned by fmod()

**Comment by apfelbox** [comment]
> The sprintf does rounding up to the 12th decimal, which means it's like your epsilon comparison, but also that is copes with the scale of the number, thanks to using the scientific notation for the rounding.

Yeah, but afterwards we still convert both to a float and do a direct comparison.
I am currently not sure whether that works out in all cases, but I trust you, if you say that this is enough. 😉 

So 👍  for the PR, as it is simpler and more readable

**Comment by nicolas-grekas** [comment]
I improved precision a bit for integers and rounded both operands to the 12th decimal for floats.

**Comment by apfelbox** [comment]
Looks great, thank you! 👍

**Comment by nicolas-grekas** [comment]
Thank you @apfelbox.

**Comment by ERuban** [comment]
@nicolas-grekas but seems that still didn't works.
As example,
fmod(0.01, 0.001) returns 0, and we got true,
but fmod(0.1, 0.001) returns 3.4694469519536E-18, and we got false.
¯\_(ツ)_/¯

**Comment by xabbuh** [comment]
@ERuban Can you please open a new issue if you think that things still need to be improved? Comments on already merged PRs are likely to get lost.

**Pull request merged**